### PR TITLE
[ISSUE 3613] bug fix, solve message hash conflict in index file

### PR DESCRIPTION
**Pull request by xijiu**
## What is the purpose of the change

bug fix, solve message hash conflict in index file

[Github issue](https://github.com/apache/rocketmq/issues/3613)

## Brief changelog

fix org.apache.rocketmq.store.DefaultMessageStore

### CODE REVIEW

File: store/src/main/java/org/apache/rocketmq/store/DefaultMessageStore.java [hunk 1]
```diff
@@ -993,6 +985,29 @@ public QueryMessageResult queryMessage(String topic, String key, int maxNum, lon
...
+     * check target msg match
+     *
+     * @param msg   message
+     * @param topic topic
+     * @param key   message key
+     * @return  is match
+     */
+    private boolean isMessageMatch(MessageExt msg, String topic, String key) {
+        String keys = msg.getKeys();
+        if (Objects.equals(topic, msg.getTopic()) && StringUtils.isNotEmpty(keys)) {
```
Comment (Git-Yang) [inline store/src/main/java/org/apache/rocketmq/store/DefaultMessageStore.java:998]: Filtering here may reduce the number of messages. It is recommended to filter in the selectPhyOffset() method.

**Comment by panzhi33** [comment]
org.apache.rocketmq.store.index.IndexService#queryOffset

File: store/src/main/java/org/apache/rocketmq/store/DefaultMessageStore.java (hunk 1 above)
Comment (xijiu) [reply store/src/main/java/org/apache/rocketmq/store/DefaultMessageStore.java:998]: filter in selectPhyOffset() may increase read commitlog times, if this is acceptable, I'll modify it

Comment (HScarb) [reply store/src/main/java/org/apache/rocketmq/store/DefaultMessageStore.java:998]: IMO, it's correct to filtering here, as we only have physical offsets in `selectPhyOffset()`, so we can't check the topic and key of message in `selectPhyOffset()`

Comment (Git-Yang) [reply store/src/main/java/org/apache/rocketmq/store/DefaultMessageStore.java:998]: Yes, I agree with you.

**Comment by HScarb** [comment]
There are some commented old code implementing topic and key check.
Anyone know Why are these code commented？

File: store/src/main/java/org/apache/rocketmq/store/DefaultMessageStore.java (hunk 1 above)
Comment (xijiu) [reply store/src/main/java/org/apache/rocketmq/store/DefaultMessageStore.java:998]: can we put check method in class `DefaultMessageStore` like this ? :

```
QueryOffsetResult queryOffsetResult = this.indexService.queryOffset(topic, key, maxNum, begin, lastQueryMsgTime, physicOffset -> {
    SelectMappedBufferResult result = commitLog.getData(physicOffset, false);
    return isTopicAndKeyMatch(result, topic, key);
});
```

if so, we will do more once IO and protocol analysis

**Comment by dongeforever** [comment]
> There are some commented old code implementing topic and key check. Anyone know Why are these code commented？ https://github.com/apache/rocketmq/pull/3616/files#diff-005813abb0e69e737cbda0fa3dcb04dddd4ab165336fae9619af60ea577fe883L958-L966
> 
> ```java
> //                    String[] keyArray = msg.getKeys().split(MessageConst.KEY_SEPARATOR);
> //                    if (topic.equals(msg.getTopic())) {
> //                        for (String k : keyArray) {
> //                            if (k.equals(key)) {
> //                                match = true;
> //                                break;
> //                            }
> //                        }
> //                    }
> ```

The logic is moved to client  in MQAdminImpl.queryMessage.
Try to reduce the load of brokers.

**Comment by dongeforever** [comment]
@xijiu Don't try to do the check in broker.

The query is usually invoked by the admin client. So it is better to do the check in client side.

If the check is done in server-side, it may affect the brokers.

**Comment by xijiu** [comment]
@dongeforever  I see. I'll correct it again

File: store/src/main/java/org/apache/rocketmq/store/DefaultMessageStore.java [hunk 2]
```diff
@@ -948,24 +952,12 @@ public QueryMessageResult queryMessage(String topic, String key, int maxNum, lon
                 long offset = queryOffsetResult.getPhyOffsets().get(m);
 
                 try {
-
-                    boolean match = true;
```
Comment (duhenglucky) [inline store/src/main/java/org/apache/rocketmq/store/DefaultMessageStore.java:952]: Actually match isn't used in broker, so how about move the check logic to client？

File: store/src/main/java/org/apache/rocketmq/store/DefaultMessageStore.java [hunk 3]
```diff
@@ -993,6 +985,29 @@ public QueryMessageResult queryMessage(String topic, String key, int maxNum, lon
...
+     * @param topic topic
+     * @param key   message key
+     * @return  is match
+     */
+    private boolean isMessageMatch(MessageExt msg, String topic, String key) {
+        String keys = msg.getKeys();
+        if (Objects.equals(topic, msg.getTopic()) && StringUtils.isNotEmpty(keys)) {
+            String[] keyArray = keys.split(MessageConst.KEY_SEPARATOR);
+            if (keyArray.length > 0) {
+                for (String k : keyArray) {
```
Comment (duhenglucky) [inline store/src/main/java/org/apache/rocketmq/store/DefaultMessageStore.java:1001]: Same as last comment

File: store/src/test/java/org/apache/rocketmq/store/index/IndexFileTest.java [hunk 4]
```diff
@@ -21,13 +21,30 @@
...
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
 
+import org.apache.rocketmq.common.BrokerConfig;
+import org.apache.rocketmq.common.Pair;
 import org.apache.rocketmq.common.UtilAll;
+import org.apache.rocketmq.common.message.MessageConst;
+import org.apache.rocketmq.common.message.MessageDecoder;
+import org.apache.rocketmq.store.*;
```
Comment (duhenglucky) [inline store/src/test/java/org/apache/rocketmq/store/index/IndexFileTest.java:38]: Please don't aggregate all imports with *

File: store/src/main/java/org/apache/rocketmq/store/DefaultMessageStore.java (hunk 2 above)
Comment (xijiu) [reply store/src/main/java/org/apache/rocketmq/store/DefaultMessageStore.java:952]: OK, I'll change it soon

**Comment by xijiu** [comment]
@dongeforever @duhenglucky     

I've moved the check logic to the client

**Review by duhenglucky — APPROVED** [review]

**Review by panzhi33 — APPROVED** [review]

**Review by vongosling — APPROVED** [review]

**Review by odbozhou — APPROVED** [review]

**Pull request merged**